File: lib/typescript/botbuilder-solutions/src/util/confirmRecognizerHelper.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ModelResult } from '@microsoft/recognizers-text';
import { recognizeBoolean } from '@microsoft/recognizers-text-choice';
import { PromptRecognizerResult } from 'botbuilder-dialogs';

export namespace ConfirmRecognizerHelper {
    const valueKey: string = 'value';

    export function confirmYerOrNo(utterance: string, locale: string = ''): PromptRecognizerResult<boolean> {
        let result: PromptRecognizerResult<boolean> = { succeeded: false };

        if (utterance !== undefined && utterance.trim().length > 0) {
            // Recognize utterance
            const results: ModelResult[] = recognizeBoolean(utterance, locale);

            if (results.length > 0) {
                const first: ModelResult = results[0];
                const value: boolean = <boolean>first.resolution[valueKey].toString();
                if (value !== undefined) {
                    result = {
                        succeeded: true,
                        value: value
                    };
                }
            }
        }

        return result;
    }
}

File: sdk/csharp/libraries/microsoft.bot.solutions/Util/ConfirmRecognizerHelper.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Recognizers.Text.Choice;

namespace Microsoft.Bot.Solutions.Util
{
    public static class ConfirmRecognizerHelper
    {
        public static PromptRecognizerResult<bool> ConfirmYesOrNo(string utterance, string locale = null)
        {
            var result = new PromptRecognizerResult<bool>();
            if (!string.IsNullOrEmpty(utterance))
            {
                // Recognize utterance
                var results = ChoiceRecognizer.RecognizeBoolean(utterance, locale);
                if (results.Count > 0)
                {
                    var first = results[0];
                    if (bool.TryParse(first.Resolution["value"].ToString(), out var value))
                    {
                        result.Succeeded = true;
                        result.Value = value;
                    }
                }
            }

            return result;
        }
    }
}
